enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,359 @@
+  `.spec.tls.externalCertificate` referencing the generated secret that was created
+  in the previous step.
+- If the secret that is referenced exists and has a successfully generated
+  cert/key pair, the router will serve this certificate if all preconditions are met.
+
+#### Variation [optional]
+
+N.A
+
+### API Extensions
+
+A `.spec.tls.externalCertificate` field is added to Route `.spec.tls` which can be used to provide a secret name
+containing the certificate data instead of using `.spec.tls.certificate` and `spec.tls.key`.
+
+```go
+
+// TLSConfig defines config used to secure a route and provide termination
+//
+// +kubebuilder:validation:XValidation:rule="has(self.termination) && has(self.insecureEdgeTerminationPolicy) ? !((self.termination=='passthrough') && (self.insecureEdgeTerminationPolicy=='Allow')) : true", message="cannot have both spec.tls.termination: passthrough and spec.tls.insecureEdgeTerminationPolicy: Allow"
+// +openshift:validation:FeatureSetAwareXValidation:featureSet=TechPreviewNoUpgrade;CustomNoUpgrade,rule="!(has(self.certificate) && has(self.externalCertificate))", message="cannot have both spec.tls.certificate and spec.tls.externalCertificate"
+type TLSConfig struct {
+	// ...
+
+	// externalCertificate provides certificate contents as a secret reference.
+	// This should be a single serving certificate, not a certificate
+	// chain. Do not include a CA certificate. The secret referenced should
+	// be present in the same namespace as that of the Route.
+	// Forbidden when `certificate` is set.
+	//
+	// +openshift:enable:FeatureSets=CustomNoUpgrade;TechPreviewNoUpgrade
+	// +optional
+	ExternalCertificate LocalObjectReference `json:"externalCertificate,omitempty" protobuf:"bytes,7,opt,name=externalCertificate"`
+}
+
+// LocalObjectReference contains enough information to let you locate the
+// referenced object inside the same namespace.
+// +structType=atomic
+type LocalObjectReference struct {
+	// name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+}
+```
+
+_Note_: The default value would be `nil`. The secret is required to be created
+in the same namespace as that of the Route. The secret must be of type
+`kubernetes.io/tls` and the tls.key and the tls.crt key must be provided in
+the `data` (or `stringData`) field of the Secret configuration.
+
+If neither `.spec.tls.externalCertificate` or `.spec.tls.certificate` and `.spec.tls.key` are
+provided the router will serve the default generated certificates.
+
+All valid and invalid scenarios will be depicted via the existing `RouteIngressCondition`.
+
+#### Variation
+
+N.A
+
+### Implementation Details/Notes/Constraints [optional]
+
+The router will read the secret referenced in `.spec.tls.externalCertificate` if present and
+if the following pre-conditions (validated in the API server) are met it uses
```

Comment:
Re-arranged this section. Should be fixed now. The conditions/validation criteria is what I was referring to here.

Reply:
Ah, did you mean the 2 validations listed below as "the following preconditions"? May be a rephrase similar to something like this will prevent the misinterpretations:

> The router will read the secret referenced in `.spec.tls.externalCertificate` and will use the certificate inside to configure HAProxy if the secret is present and if the following pre-conditions are met: